<div class="container">

<table style="width: 100%;"><tr>
<td>xegaGeGene</td>
<td style="text-align: right;">R Documentation</td>
</tr></table>
<h2>Package xegaGeGene.</h2>

<h3>Description</h3>

<p>The <code>xegaGeGene</code> package 
provides functions implementing
grammatical evolution with binary coded genes:
</p>


<h3>Details</h3>


<ul>
<li>
<p> Gene initialization.
</p>
</li>
<li>
<p> Gene maps for the mod and (approximately) for the bucket rule.
</p>
</li>
<li>
<p> Grammar-based decoders for binary coded genes.
</p>
</li>
<li>
<p> Analysis of the interaction of  codon precision 
with the rule choice bias for a given grammar.
</p>
</li>
<li>
<p> Automatic determination of codon precision 
with a limited rule choice bias.
</p>
</li>
</ul>
<h3>Gene Initialization</h3>

<p>The number of bits of a gene are specified by <code>lF$BitsOnGene()</code>.
</p>
<p>The number of bits of a codon are specified 
by <code>lF$CodonPrecision()</code>.
</p>


<h3>Binary Gene Representation</h3>

<p>A binary gene is a named list:
</p>

<ul>
<li>
<p> $gene1      the gene must be a binary vector.
</p>
</li>
<li>
<p> $fit        the fitness value of the gene
(for EvalGeneDet and EvalGeneU) or
the mean fitness (for stochastic functions
evaluated with EvalGeneStoch).
</p>
</li>
<li>
<p> $evaluated  has the gene been evaluated?
</p>
</li>
<li>
<p> $evalFail   has the evaluation of the gene failed?
</p>
</li>
<li>
<p> $var        the cumulative variance of the fitness 
of all evaluations of a gene.
(For stochastic functions)
</p>
</li>
<li>
<p> $sigma      the standard deviation of the fitness of 
all evaluations of a gene.
(For stochastic functions)
</p>
</li>
<li>
<p> $obs        the number evaluations of a gene.
(For stochastic functions)
</p>
</li>
</ul>
<h3>Abstract Interface of Problem Environment</h3>

<p>A problem environment <code>penv</code> must provide:
</p>

<ul>
<li> <p><code>$f(parameters, gene, lF)</code>: 
Function with a real parameter vector as first argument 
which returns a gene 
with evaluated fitness.
</p>
</li>
<li>
<p> $genelength(): The number of bits of the binary coded
real parameter vector. Used in <code>InitGene</code>.
</p>
</li>
<li>
<p> $bitlength(): A vector specifying the number of bits 
used for coding each real parameter.
If <code>penv$bitlength()[1]</code> is <code>20</code>, 
then <code>parameters[1]</code> is coded by 20 bits.
Used in <code>GeneMap</code>.
</p>
</li>
<li>
<p> $lb(): The lower bound vector of each parameter.
Used in <code>GeneMap</code>.
</p>
</li>
<li>
<p> $ub(): The upper bound vector of each parameter.
Used in <code>GeneMap</code>.
</p>
</li>
</ul>
<h3>The Architecture of the xegaX-Packages</h3>

<p>The xegaX-packages are a family of R-packages which implement 
eXtended Evolutionary and Genetic Algorithms (xega).  
The architecture has 3 layers, 
namely the user interface layer,
the population layer, and the gene layer: 
</p>

<ul>
<li>
<p>The user interface layer (package <code>xega</code>) 
provides a function call interface and configuration support
for several algorithms: genetic algorithms (sga), 
permutation-based genetic algorithms (sgPerm), 
derivation free algorithms as e.g. differential evolution (sgde), 
grammar-based genetic programming (sgp) and grammatical evolution
(sge). 
</p>
</li>
<li>
<p>The population layer (package <code>xegaPopulation</code>) contains
population related functionality as well as support for 
population statistics dependent adaptive mechanisms and parallelization.
</p>
</li>
<li> 
<p>The gene layer is split in a representation independent and 
a representation dependent part:
</p>

<ol>
<li> 
<p>The representation indendent part (package <code>xegaSelectGene</code>)
is responsible for variants of selection operators, evaluation 
strategies for genes, as well as profiling and timing capabilities.        
</p>
</li>
<li> 
<p>The representation dependent part consists of the following packages: 
</p>

<ul>
<li> <p><code>xegaGaGene</code> for binary coded genetic algorithms.
</p>
</li>
<li> <p><code>xegaPermGene</code> for permutation-based genetic algorithms.
</p>
</li>
<li> <p><code>xegaDfGene</code> for derivation free algorithms as e.g. 
differential evolution.
</p>
</li>
<li> <p><code>xegaGpGene</code> for grammar-based genetic algorithms.
</p>
</li>
<li> <p><code>xegaGeGene</code> for grammatical evolution algorithms.
</p>
</li>
</ul>
<p>The packages <code>xegaDerivationTrees</code> and <code>xegaBNF</code> support
the last two packages:
<code>xegaBNF</code> essentially provides a grammar compiler and 
<code>xegaDerivationTrees</code> an abstract data type for derivation trees.
</p>
</li>
</ol>
</li>
</ul>
<h3>Copyright</h3>

<p>(c) 2024 Andreas Geyer-Schulz
</p>


<h3>License</h3>

<p>MIT
</p>


<h3>URL</h3>

<p>https://github.com/ageyerschulz/xegaGeGene
</p>


<h3>Installation</h3>

<p>From CRAN by <code>install.packages('xegaGeGene')</code>
</p>


<h3>Author(s)</h3>

<p>Andreas Geyer-Schulz
</p>


<h3>References</h3>

<p>Ryan, Conor and Collins, J. J. AND Neill, Michael O. (1998)
Grammatical evolution: Evolving programs for an arbitrary language.
In: 	Banzhaf, Wolfgang and Poli, Riccardo, Schoenauer, Marc and 
Fogarty, Terence C. (1998):
Genetic Programming. First European Workshop, EuroGP' 98 
Paris, France, April 14-15, 1998 Proceedings,
Lecture Notes in Computer Science, 1391, Springer, Heidelberg.
&lt;doi:10.1007/BFb0055930&gt;
</p>
<p>O'Neil, Michael AND Ryan, Conor (2003)
Grammatical Evolution: 
Evolutionary Automatic Programming in an Arbitrary Language.
Kluwer, Dordrecht.
&lt;ISBN:1-4020-7444-1&gt;
</p>
<p>Ryan, Conor and O'Neill, Michael and Collins, J. J. (2018)
Handbook of Grammatical Evolution.
Springer International Publishing, Cham.
&lt;doi:10.1007/978-3-319-78717-6&gt;
</p>


</div>